fl4health/clients/mr_mtl_client.py

```diff
@@ -0,0 +1,188 @@
+                model (global model for all but the first step of MR-MTL)
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+            fitting_round (bool): Boolean that indicates whether the current federated learning
+                round is a fitting round or an evaluation round.
+                This is used to help determine which parameter exchange should be used for pulling parameters.
+                A full parameter exchanger is only used if the current federated learning round is the very
+                first fitting round.
+        """
+        # Make sure that the proper components exist.
+        assert self.init_global_model is not None and self.model is not None
+        assert self.parameter_exchanger is not None and isinstance(self.parameter_exchanger, FullParameterExchanger)
+
+        # Route the parameters to the GLOBAL model in MR-MTL
+        assert self.parameter_exchanger is not None
+        log(INFO, "Setting the global model weights")
+        self.parameter_exchanger.pull_parameters(parameters, self.init_global_model, config)
+
+    def update_before_train(self, current_server_round: int) -> None:
+        assert isinstance(self.model, nn.Module)
+        # Freeze the initial weights GLOBAL MODEL. These are used to form the MR-MTL
+        # update penalty term.
+        for param in self.init_global_model.parameters():
+            param.requires_grad = False
+        self.init_global_model.eval()
+
+        return super().update_before_train(current_server_round)
+
+    def compute_training_loss(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> TrainingLosses:
+        """
+        Computes training losses given predictions of the modes and ground truth data. We add to vanilla loss
+        function by including Mean Regularized (MR) penalty loss which is the l2 inner product between the
+        initial global model weights and weights of the current model.
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name.
+                All predictions included in dictionary will be used to compute metrics.
+            features: (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target: (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            TrainingLosses: an instance of TrainingLosses containing backward loss and
+                additional losses indexed by name. Additional losses includes each loss component of the total loss.
+        """
+        # Check that both models are in training mode
+        assert not self.init_global_model.training and self.model.training
+
+        total_loss, additional_losses = self.compute_loss_and_additional_losses(preds, features, target)
+        if additional_losses is None:
```

Comment:
TrainingLosses actually handles the setting where `additional_losses` is None. So no need to do this step 🙂